src/cockpit/peer.py

```diff
@@ -161,6 +163,11 @@ def do_closed(self, exc: Optional[Exception]) -> None:
     def process_exited(self) -> None:
         assert isinstance(self.transport, SubprocessTransport)
         logger.debug('Peer %s exited, status %d', self.__class__.__name__, self.transport.get_returncode())
+        # a common case is that the called peer does not exist
+        if self.transport.get_returncode() == 127:
+            self.close(PeerError('no-cockpit', error='no-cockpit'))
+        else:
+            self.close(EOFError(f'Peer exited with status {self.transport.get_returncode()}'))
```

Comment:
`EOFError` seems weird here since this isn't in response to EOF.  I know that we need to do this, though, to get the detailed error messages out of ferny in case of a generic shutdown that we don't have better information about.

What if we made something like a `PeerTerminatedError` subclass of `PeerError` which defaults the problem code to `'terminated'` and raised that instead of EOFError here?  That would give us something unique to catch and would also mean that we don't need to translate `EOFError` into `problem='terminated'` later on.

Reply:
I'm not sure how to translate this -- superuser.py does *not* catch EOFError, it just catches a few specific ones in `go()`. It does catch `PeerError`, which is precisely what makes the test case fail. So it would catch a subclass as well. 

I don't like this either, it's strange. I'll look at this again tmw.

Reply:
I reworked this a bit.  The problem with the patch above was that superuser.py does not expect an exception at all, just a close message.

I'm not 100% sure that it's right, but at least it avoids this strange EOFError conversion.

Reply:
Dang.. See below -- [that attempted cleanup](https://github.com/cockpit-project/cockpit/compare/d0187f66993bd7ed0e68bf97fcd879d095c4a908..b3486bc8eb5bfdf37dab15d5fe6f8b8e2688ada7) causes [two regressions](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18864-20230601-050823-b3486bc8-fedora-38-pybridge/log.html). So I reverted this for now -- I'm afraid I don't fully understand the interactions here.